pkg/provisioningv2/rke2/planner/encryptionkeyrotation.go

```diff
@@ -104,29 +155,142 @@ func hasValidRotateEncryptionKeyStatus(cp *rkev1.RKEControlPlane) bool {
+				return entry, nil
+			}
+		}
+	}
+
+	machines := collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)
+	if len(machines) == 0 {
+		return nil, p.encryptionKeyRotationFailed(cp)
+	}
+	if cp.Status.RotateEncryptionKeysPhase != rkev1.RotateEncryptionKeysPhaseStart {
+		return nil, p.restartEncryptionKeyRotation(cp)
+	}
+
+	cp.Annotations[encryptionKeyRotationLeaderAnnotation] = machines[0].Machine.Name
+	_, err := p.rkeControlPlanes.Update(cp)
+	if err != nil {
+		return nil, err
+	}
+	// a new leader was chosen, must restart
+	return nil, generic.ErrSkip
+}
+
+// encryptionKeyRotationIsSuitableControlPlane ensures that a control plane node has not been deleted and has a valid
+// node associated with it.
+func encryptionKeyRotationIsSuitableControlPlane(entry *planEntry) bool {
+	return isControlPlane(entry) && entry.Machine.DeletionTimestamp == nil && entry.Machine.Status.NodeRef != nil
+}
+
+// encryptionKeyRotationStart gets the current status of the leader before allowing a restart to happen. Although this
+// plan should be close to in sync on subsequent iterations, it is necessary since secrets-encrypt commands can be run
+// on followers, which would cause the leader node to update its status, however the followers plans would be in sync
+// which could result in a deadlock.
 func (p *Planner) encryptionKeyRotationStart(cp *rkev1.RKEControlPlane, leader *planEntry) error {
 	return p.encryptionKeyRotationGetLeaderStatus(cp, leader, rkev1.RotateEncryptionKeysPhaseRestartLeader)
 }
 
-func (p *Planner) encryptionKeyRotationRestartLeader(cp *rkev1.RKEControlPlane, leader *planEntry) error {
-	phase, err := encryptionKeyRotationPhaseFromPeriodic(leader)
+// encryptionKeyRotationRestartLeader will restart the leader to ensure that the current encryption key rotation status
+// is persisted to the datastore. With a single node control plane, the apiserver container needs to be restarted,
+// however with an HA control plane, the leader must first be restarted then all the followers, as the followers will
+// update their status when reading from the datastore on startup of the rke2/k3s services.
+func (p *Planner) encryptionKeyRotationRestartLeader(cp *rkev1.RKEControlPlane, clusterPlan *plan.Plan, leader *planEntry) error {
+	status, err := encryptionKeyRotationSecretsEncryptStatusFromPeriodic(leader)
 	if err != nil {
 		return p.enqueueIfErrWaiting(cp, err)
 	}
-	nodePlan := plan.NodePlan{
-		Instructions: []plan.OneTimeInstruction{encryptionKeyRotationRestartInstruction(cp, leader, phase)},
+
+	nodes := collect(clusterPlan, encryptionKeyRotationIsSuitableControlPlane)
+	if len(nodes) == 0 {
+		return p.restartEncryptionKeyRotation(cp) // actually should bail out here, not even a leader
 	}
-	err = assignAndCheckPlan(p.store, fmt.Sprintf("encryption key rotation: [%s]", cp.Status.RotateEncryptionKeysPhase), leader, nodePlan, 0)
-	if err != nil {
-		return p.enqueueIfErrWaiting(cp, err)
+	if len(nodes) == 1 {
```

Comment:
Does the fix in RKE2 mean we don't have to handle single control-plane nodes differently?

Reply:
We still have to treat single control-plane nodes different (restarting kube-apiserver container for single node, and `systemctl restart rke2-server` for multiple). The fix on the rke2 side was for control-plane nodes that did not also have the etcd role.